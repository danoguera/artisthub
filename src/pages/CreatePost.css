.service{
  display: flex;
  justify-content: center;
  margin: 10px;
  text-align: center;
  color: #FFFFFF;
  }
  
  .boxService{
    background-color: #E8E8E8;
    width: 450px;
    height: 700px;
    /* margin: auto; */
    background: rgba(0,0,0,0.4);
    padding: 10px;
    border-radius: 8px;
  }
  
  .boxService h1{
    font-size: 30px;
    padding-bottom: 10px;
  }
  
  .boxService h2{
    font-size: 20px;
  }
  
  .boxService select{
    font-size: 18px;
    display: block;
    background: rgba(240,240,240,0.05);
    color: #FFFFFF;
    padding: 10px 15px;
    width: 89%;
    max-width: 100%;
    /* margin: 20px; */
    border-radius: 5px;
    -moz-appearance: none;
    -webkit-appearance: none;
    appearance: none;
    line-height: 20px;
    border: 1px solid rgb(199, 190, 190);
    background-image: url('../assets/images/down-arrow.png');
    background-repeat: no-repeat, repeat;
    background-position: right .7em top 50%;
    background-size: .90em;
    margin: 0 auto;
  }
  
  .boxService select option {
    color: black;
  }

  input[type="file"]{
    display: none;
  }
  
  .boxService label{
    display: flex;
    justify-content: center;
    align-items: center;
    background: rgba(240,240,240,0.05);
    border: 1px solid rgb(199, 190, 190);
    border-radius: 5px;
    line-height: 2px;
    /* margin-top: 10px; */
    margin: 0 auto;
    width: 89%;
    padding: 3px 0;
  }
  
  .boxService label p{
    margin-left: 6px;
    font-size: 18px;
  }
  
  ::placeholder{
    color: #FFFFFF;
    font-size: 18px;
  }
  
  .inputBoxService{
    display: block;
    box-sizing: border-box;
    margin: auto;
    padding: 15px 15px;
    background: rgba(240,240,240,0.05);
    width: 89%;
    height: 40px;
    border: 1px solid rgb(199, 190, 190);
    outline: none;
    font-size: 18px;
    transition: 0.2s;
    color: #FFFFFF;
    border-radius: 5px;
  }
  
  .textBox{
    display: block;
    box-sizing: border-box;
    margin: auto;
    padding: 10px 15px;
    background: transparent;
    width: 89%;
    height: 92px;
    border: 1px solid rgb(199, 190, 190);
    outline: none;
    font-size: 15px;
    transition: 0.2s;
    color: #FFF;
    /* text-shadow: 2px 2px 8px #fff; */
    font-weight: 400;
    border-radius: 5px;
  }
  
  .submit-btn{
    width: 120px;
    height: 32px;
    background: #F44336;
    border: none;
    border-radius: 8px;
    color: #FFF;
    cursor: pointer;
    transition: 0.2s;
    font-size: 20px;
  }
  
  @media only screen and (min-width: 1000px){

    .service{
      position: absolute;
      /* margin-top: 70px; */
      text-align: center;
      color: #FFFFFF;
      line-height: 20px;
      top: 35%;
      left: 50%;
      transform: translate(-50%,-50%);
      height: 500px;
    }
  
    .service{
      height: 400px;
      margin-top: 90px;
    }
  
    .boxService{
      background-color: #E8E8E8;
      width: 1000px;
      height: 750px;
      /* margin: auto; */
      background: rgba(0,0,0,0.4);
      padding: 10px;
      border-radius: 8px;
      padding: 20px 0;
      margin-bottom: 10px;
    }
    
    .boxService h1{
      font-size: 30px;
    }
    
    .boxService h2{
      font-size: 20px;
    }


    input[type="file"]{
      display: none;
    }
    
    .boxService label{
      display: flex;
      justify-content: center;
      align-items: center;
      background: rgba(240,240,240,0.05);
      border: 1px solid rgb(199, 190, 190);
      border-radius: 6px;
      line-height: 2px;
      margin: auto;
      width: 89%;
    }
    
    .boxService label p{
      margin-left: 6px;
      font-size: 18px;
    }
    
    ::placeholder{
      color: #FFFFFF;
    }
/*     
    .inputBoxService{
      display: block;
      box-sizing: border-box;
      margin: auto;
      padding: 10px;
      background: rgba(240,240,240,0.05);
      width: 89%;
      height: 32px;
      border: 1px solid rgb(199, 190, 190);
      outline: none;
      font-size: 15px;
      transition: 0.2s;
      color: #FFFFFF;
      border-radius: 5px;
    } */
    
    /* .textBox{
      display: block;
      box-sizing: border-box;
      margin: auto;
      padding: 10px;
      background: transparent;
      width: 89%;
      height: 92px;
      border: 1px solid rgb(199, 190, 190);
      outline: none;
      font-size: 15px;
      transition: 0.2s;
      color: #FFF;
      font-weight: 400;
      border-radius: 5px;
    } */
  
  }