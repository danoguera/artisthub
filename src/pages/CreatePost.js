import React from 'react';
import axios from 'axios';

class CreatePost extends React.Component{
    constructor(){
        super();
        this.state = {
            email: "",
            password: "",
            result: "",
            subcategoryList: [] ,
            subcategory: "",
            category: "",
            country: "",
            state: "",
            city: "",
            title: "", 
            description: "", 
            post_image: "",

        } 
    } 

    handleSelect = (event) => {
        console.log("Name:",event.target.name, "Valor:",event.target.value)
        this.setState({
           [event.target.name]: event.target.value 
        })
        if (event.target.value="photography"){
            this.setState({
                subcategoryList:["aerial","model","food","wedding"]  
            })
         } 
    } 

    handleSubmit = (event) => {
        event.preventDefault();
        const {title, description, city, state, country, subcategory, category, post_image} = this.state; 
        axios({
            url: process.env.REACT_APP_SERVER_URL+"/posts/",
            method: "POST",
            headers: { "Authorization": localStorage.getItem("token") },
            data: {
                title,
                description,
                city,
                state, 
                country,
                subcategory,
                category,
                post_image,
            }
            })
            .then(response => {
                console.log(response.data);
                //localStorage.setItem("token",response.data);
                //this.props.onUpdate(response.data);
                this.props.history.push("/posts/"+response.data._id);  
            })
            .catch(error =>{
                console.log(error);
                 this.setState({ error: error,
                    password: "",
                })
             })
            .finally(() => this.setState({ loading: false })); 
    
    } 

    handleInput = (event) => {
        this.setState({
           [event.target.name]: event.target.value, 
        })    
    } 

    render(){
        if (this.state.loading){return <h1>Loading...</h1>}
      

        return (
            <main>
                <form onSubmit={this.handleSubmit} >
                    <h1>Create your service</h1>
                    <h2>Please fill the form...</h2>
                    <label htmlFor="title" >Title: </label>
                    <input id="title" name="title" type="text" className="inputBox" placeholder="Title..." value={this.state.title} onChange={this.handleInput} />
                    <label htmlFor="description"> Description: </label>
                    <textarea id="description" name="description" className="inputBox" placeholder="Describe your service ..." value={this.state.description} onChange={this.handleInput} />
                    <label htmlFor="category">Choose a service:</label>
                    <select name="category" onChange={this.handleSelect} id="category">
                        <option value="music">Music</option>
                        <option value="photography">Photography</option>
                        <option value="videography">Videography</option>
                    </select>
                    <br />

                    <label htmlFor="subcategory">Choose a subcategory:</label>
                    <select value={this.state.subcategory} name="subcategory" id="subcategory" onChange={this.handleSelect} >
                        {this.state.subcategoryList.map(option => (
                            <option value={option} > {option.charAt(0).toUpperCase() + option.slice(1)} </option>
                        )
                        )};
                </select>
                <br />

                    <label htmlFor="post_image" >Title: </label>
                    <input id="post_image" name="post_image" type="text" className="inputBox" placeholder="Ruta a la foto..." value={this.state.post_image} onChange={this.handleInput} />
                    <br />

                    <select onChange={this.handleInput} name="country" class="countries" id="countryId">
                        <option value="">Select Country</option>
                        <option value="Australia">Australia </option>
                        <option value="Colombia">Colombia</option>
                    </select>
                    <select onChange={this.handleInput} name="state" class="states" id="stateId">
                        <option value="">Select State</option>
                        <option value="Victoria">Victoria</option>
                        <option value="Atlantico">Atlantico</option>
                        <option value="Bogota">Bogota</option>
                    </select>
                    <select onChange={this.handleInput} name="city" class="cities" id="cityId">
                        <option value="Bogota">Bogota</option>
                        <option value="Barranquilla">Curramba</option>
                        <option value="Melbourne">Melbourne</option>
                    </select>
                    <br />

                    <input type="submit" onSubmit={this.handleSubmit} className="submit-btn" placeholder="Sign in" value="Submit" />
                </form>
            </main>
        )

    }
} 

export default CreatePost;
