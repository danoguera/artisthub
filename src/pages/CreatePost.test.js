import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { render, cleanup,  fireEvent, waitForElement } from '@testing-library/react';
import moxios from 'moxios';
import CreatePost from './CreatePost';

describe('CreatePost', () => {
  beforeEach(() => moxios.install());
  
  afterEach(() => {
    moxios.uninstall();
    cleanup();
  });

  it('should render correctly', () => {
    const match = { params: { } };
    const { container } = render(<CreatePost match={match} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should change data as user writes in the input boxes ', (done) => {
    const history = [];
    const match = { params: { } };
    const { getByTestId, getByPlaceholderText } = render(
      <MemoryRouter>
        <CreatePost match={match} history={history} />
      </MemoryRouter>
    );

    const input = getByPlaceholderText('Enter your service title');
    const event = {
      target: {
        name: 'title',
        value: 'Description 1',
      },
    };

    fireEvent.change(input, event);   //Hasta aqui simulamos escribir una clave

    expect(getByTestId('post-title').value).toBe("Description 1");
    done();
  });

  it('should change data if user selects from dropdown ', (done) => {
    const history = [];
    const match = { params: { } };
    const { getByTestId, getByPlaceholderText } = render(
      <MemoryRouter>
        <CreatePost match={match} history={history} />
      </MemoryRouter>
    );

    const input = getByTestId('category');
    const event = {
      target: {
        name: 'category',
        value: 'Videography',
      },
    };

    fireEvent.change(input, event);   //Hasta aqui simulamos escribir una clave

    expect(getByTestId('category').value).toBe("Videography");
    done();
  });


  it('should create post ', (done) => {
    const history = [];
    const match = { params: { } };
    const { getByTestId, getByPlaceholderText } = render(
      <MemoryRouter>
        <CreatePost match={match} history={history} />
      </MemoryRouter>
    );

    const input = getByPlaceholderText('Enter your service title');
    const event = {
      target: {
        name: 'title',
        value: 'Description 1',
      },
    };

    fireEvent.change(input, event); 

    expect(getByTestId('post-title').value).toBe("Description 1");

    fireEvent.click(getByTestId('submit-btn'));

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
       status: 200,
       response: {_id:1111, title:"Post 1", description:"Post 1 description", post_image:"http://loquesea.com/image.jpg"}
        }).then(() => {
       
       expect(history[0]).toMatch("/posts/1111"); 
       done();
      })
    });

  });


  it('should display an alert if an error occurs while creating a post ', (done) => {
    window.alert = jest.fn();
    const history = [];
    const match = { params: { } };
    const { getByTestId, getByPlaceholderText } = render(
      <MemoryRouter>
        <CreatePost match={match} history={history} />
      </MemoryRouter>
    );

    const input = getByPlaceholderText('Enter your service title');
    const event = {
      target: {
        name: 'title',
        value: 'Description 1',
      },
    };

    fireEvent.change(input, event); 

    expect(getByTestId('post-title').value).toBe("Description 1");

    fireEvent.click(getByTestId('submit-btn'));

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
       status: 404,
       response: {}
        }).then(() => {
          //No hay mucho que porbar aca, si se genera error, solo hay un alertBox.. todo sigue igual
          expect(getByTestId('post-title').value).toBe("Description 1");
       done();
      })
    });

  });

  it('should render the form with the data of an existing post ', (done) => {
    const history = [];
    const match = { params: { id: 1111} };
    const { getByTestId, getByPlaceholderText } = render(
      <MemoryRouter>
        <CreatePost match={match} history={history} />
      </MemoryRouter>
    );

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
       status: 200,
       response: {_id:1111, title:"Post 1", description:"Post 1 description", post_image:"http://loquesea.com/image.jpg"}
        }).then(() => {
       
       expect(getByTestId('post-title').value).toBe("Post 1");
       done();
      })
    });

    

  });


  it('should create post with image ', (done) => {
    const history = [];
    const match = { params: { } };
    const { getByTestId, getByPlaceholderText } = render(
      <MemoryRouter>
        <CreatePost match={match} history={history} />
      </MemoryRouter>
    );

    const file = new File(['(Â¬?_?)'], 'chucknorris.png', { type: 'image/png' })
    const imageInput = getByTestId("post_image")
    fireEvent.change(imageInput, { target: { files: [file] } })

    fireEvent.click(getByTestId('submit-btn'));

    moxios.wait(() => {
      const request = moxios.requests.mostRecent();
      request.respondWith({
       status: 200,
       response: {_id:1111, title:"Post 1", description:"Post 1 description", post_image:"http://loquesea.com/image.jpg"}
        }).then(() => {
       
       expect(history[0]).toMatch("/posts/1111"); 
       done();
      })
    });

  });




})