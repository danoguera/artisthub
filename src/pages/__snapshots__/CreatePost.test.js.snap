// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePost should render correctly 1`] = `
<main>
  <section>
    <div
      class="service"
    >
      <div
        class="boxService"
      >
        <h1>
          Create your service
        </h1>
        <h2>
          Please fill up the form
        </h2>
        <form>
          <input
            class="inputBoxService"
            data-testid="post-title"
            id="title"
            name="title"
            placeholder="Enter your service title"
            type="text"
            value=""
          />
          <br />
          <textarea
            class="textBox"
            id="description"
            name="description"
            placeholder="Enter your service description"
          />
          <br />
          <select
            data-testid="category"
            id="category"
            name="category"
          >
            <option
              value=""
            >
              Pick your Service:
            </option>
            <option
              value="Music"
            >
              Music
            </option>
            <option
              value="Photography"
            >
              Photography
            </option>
            <option
              value="Videography"
            >
              Videography
            </option>
          </select>
          <br />
          <select
            id="subcategory"
            name="subcategory"
          >
            <option
              value=""
            >
              Select a Subcategory
            </option>
            <option
              value="Wedding"
            >
              Wedding
            </option>
            <option
              value="Aerial"
            >
              Aerial
            </option>
            <option
              value="Food"
            >
              Food
            </option>
            <option
              value="Models"
            >
              Models
            </option>
          </select>
          <br />
          <select
            class="countries"
            id="countryId"
            name="country"
          >
            <option
              value=""
            >
              Select your country:
            </option>
            <option
              value="Australia"
            >
              Australia 
            </option>
            <option
              value="Colombia"
            >
              Colombia
            </option>
            <option
              value="Congo"
            >
              Congo
            </option>
            <option
              value="Croatia"
            >
              Croatia
            </option>
            <option
              value="Cyprus"
            >
              Cyprus
            </option>
          </select>
          <br />
          <select
            class="states"
            id="stateId"
            name="state"
          >
            <option
              value=""
            >
              Select your state:
            </option>
            <option
              value="Antioquia"
            >
              Antioquia
            </option>
            <option
              value="Atlantico"
            >
              Atlantico
            </option>
            <option
              value="Bogota"
            >
              Bogota
            </option>
            <option
              value="Santander"
            >
              Santander
            </option>
            <option
              value="Victoria"
            >
              Victoria
            </option>
          </select>
          <br />
          <select
            class="cities"
            id="cityId"
            name="city"
          >
            <option
              value=""
            >
              Select your city:
            </option>
            <option
              value="Bogota"
            >
              Bogota
            </option>
            <option
              value="Barranquilla"
            >
              Curramba
            </option>
            <option
              value="Bucaramanga"
            >
              Bucaramanga
            </option>
            <option
              value="Cali"
            >
              Cali
            </option>
            <option
              value="Medellin"
            >
              Medellin
            </option>
            <option
              value="Melbourne"
            >
              Melbourne
            </option>
          </select>
          <br />
          <input
            data-testid="post_image"
            id="post_image"
            name="post_image"
            type="file"
            value=""
          />
          <label
            for="post_image"
          >
            <span
              class="material-icons"
            >
              cloud_upload
            </span>
            <p>
              Upload a Photo
            </p>
          </label>
          <br />
          <input
            class="submit-btn"
            data-testid="submit-btn"
            type="submit"
            value="Submit"
          />
        </form>
      </div>
    </div>
  </section>
</main>
`;
